File: app/src/main/java/workshop/akbolatss/tagsnews/screen/board/SectionsPagerAdapter.java
package workshop.akbolatss.tagsnews.screen.board;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import java.util.List;

import workshop.akbolatss.tagsnews.screen.news.NewsSource;
import workshop.akbolatss.tagsnews.util.SmartFragmentStatePagerAdapter;


public class SectionsPagerAdapter extends SmartFragmentStatePagerAdapter<Fragment> {

    private List<NewsSource> sections;

    public SectionsPagerAdapter(FragmentManager fm, List<NewsSource> sections) {
        super(fm);
        this.sections = sections;
    }

    @Override
    public Fragment getItem(int position) {
        Bundle bundle = new Bundle();
        bundle.putString(String.valueOf(position), sections.get(position).getName());
        bundle.putString(getPageTitle(position).toString(), sections.get(position).getUrl());
        sections.get(position).getFragment().setArguments(bundle);
        return sections.get(position).getFragment();
    }

    @Override
    public int getCount() {
        if (sections == null) {
            return 0;
        } else {
            return sections.size();
        }
    }


    public void onUpdate(List<NewsSource> newSections) {
        for (int i = getCount() - 1; i >= 1; i--) {
            sections.remove(i);
        }
        sections.addAll(newSections);
        notifyDataSetChanged();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return sections.get(position).getName();
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/board/SectionsPagerAdapter.kt
package workshop.akbolatss.tagsnews.screen.board

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import workshop.akbolatss.tagsnews.model.dao.RssSource
import workshop.akbolatss.tagsnews.screen.news.NewsFragment
import workshop.akbolatss.tagsnews.screen.recommendations.RecommendationsFragment
import workshop.akbolatss.tagsnews.util.SmartFragmentStatePagerAdapter

/**
 * FragmentStatePager adapter to handle #RecommendationsFragment and #NewsFragment in #BoardActivity
 * @see BoardActivity
 */
class SectionsPagerAdapter(fm: FragmentManager, private val sections: MutableList<RssSource>) : SmartFragmentStatePagerAdapter<Fragment>(fm) {

    override fun getItem(position: Int): Fragment {
        return if (position == 0) {
            RecommendationsFragment()
        } else {
            NewsFragment.newInstance(sections[position - 1].id, sections[position - 1].link)
        }
    }

    override fun getCount(): Int {
        return sections.size + 1
    }

    fun onUpdate(newSections: List<RssSource>) {
        if (count > 1) {
            for (i in sections.size downTo 1) {
                sections.removeAt(i - 1)
            }
        }
        sections.addAll(newSections)
        notifyDataSetChanged()
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return if (position == 0) {
            "+"
        } else {
            sections[position - 1].title
        }
    }
}
